packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,622 @@
  *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
  */
 
+import { Model, transaction } from 'objection';
+import { handleObjectionError } from '../../util/errorCodes.js';
+import { oneExists, oneTruthy, setFalsyValuesToNull } from '../../util/middleware.js';
+import {
+  customError,
+  checkIsArray,
+  checkIdIsNumber,
+  checkExactlyOneIsProvided,
+  checkRecordBelongsToFarm,
+} from '../../util/customErrors.js';
+
 import AnimalModel from '../../models/animalModel.js';
 import AnimalBatchModel from '../../models/animalBatchModel.js';
+import CustomAnimalTypeModel from '../../models/customAnimalTypeModel.js';
+import DefaultAnimalBreedModel from '../../models/defaultAnimalBreedModel.js';
+import CustomAnimalBreedModel from '../../models/customAnimalBreedModel.js';
+import AnimalUseModel from '../../models/animalUseModel.js';
+import AnimalOriginModel from '../../models/animalOriginModel.js';
+import AnimalIdentifierType from '../../models/animalIdentifierTypeModel.js';
 
 const AnimalOrBatchModel = {
   animal: AnimalModel,
   batch: AnimalBatchModel,
 };
 
-export function checkEditAnimalOrBatch(animalOrBatchKey) {
+const checkValidAnimalOrBatchIds = async (animalOrBatchKey, ids, farm_id, trx) => {
+  if (!ids || !ids.length) {
+    throw customError('Must send ids');
+  }
+
+  const idsSet = new Set(ids.split(','));
+
+  // Check that all animals/batches exist and belong to the farm
+  const invalidIds = [];
+
+  for (const id of idsSet) {
+    // For query syntax like ids=,,, which will pass the above check
+    checkIdIsNumber(id);
+
+    const existingRecord = await AnimalOrBatchModel[animalOrBatchKey]
+      .query(trx)
+      .findById(id)
+      .where({ farm_id })
+      .whereNotDeleted(); // prohibiting re-delete
+
+    if (!existingRecord) {
+      invalidIds.push(id);
+    }
+  }
+
+  if (invalidIds.length) {
+    throw customError(
+      'Some entities do not exist, are already deleted, or are not associated with the given farm.',
+      400,
+      { error: 'Invalid ids', invalidIds },
+    );
+  }
+};
+
+// For edit mode set creating to false
+const checkAnimalType = async (animalOrBatch, farm_id, creating = true) => {
+  const { default_type_id, custom_type_id, type_name } = animalOrBatch;
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+  // Skip if all undefined or editing (!creating)
+  if (creating || oneTruthy([default_type_id, custom_type_id, type_name])) {
```

Comment:
In the editing case:  if `oneTruthy()` then it will be checked. If `!oneTruthy()` it doesn't need to be checked because it is not required on edit. 

But I may have missed a case 
```
if (oneExists && !oneTruthy) { throw 'Cannot delete animal type'}
```
Or something like that

Reply:
Resolving this since zero of the values should be able to be present while editing, and the case I suggested might be a problem is handled with a db error. I forget where we landed regarding early and named error returns - but I like plain check violation errors as it is less code.